src/metabase/fingerprinting.clj

```diff
@@ -177,15 +177,23 @@
                            (- 1 cardinality-error))
         :entropy (binned-entropy histogram)}))))
 
-(defmulti fingerprint class)
+(defmulti fingerprint (fn [_ x] (class x)))
+
+(def ^:private ^:const max-sample-size 10000)
+
+(defn- extract-query-opts
+  [{:keys [max-cost]}]
+  (cond-> {}
+    (< (or max-cost 5) 3) (assoc :limit max-sample-size)))
```

Comment:
We should use some sort of a meaningful constant name for this, rather than magic numbers.